### PR TITLE
dynamic estimation of baro_gps_offset

### PR DESCRIPTION
**Pull request by kd0aij**
use low pass filtered baro and gps altitudes to maintain a dynamic offset
reduce baro temp sample rate to ~1Hz
baro pressure sample rate is ~72Hz

### CODE REVIEW

**Comment by kd0aij** [comment]
Finally got to fly with this active. It looks to me as though the offset estimation is working: GPOS.alt tracks GPS.alt closely but shows high frequency info coming from the barometer.
http://dash.oznet.ch/view/hfwtpvuMWDpHbTRCxh9B2M

The code is still stuffing GPOS.TALT, eph, epv with offset, filtered baro alt and filtered gps alt, and here's what the filtered altitudes look like:
baro_offset_flight_test

If nobody sees any problems with this, I'll remove the debug code and commit that change here.

**Comment by LorenzMeier** [comment]
Awesome! With the debugging log stuff reset to default this looks like it should go in.

Upon looking through your log I noticed that you have a constant offset of the pitch rate - you might want to give the pitch rate controller a bigger I term to compensate for that. The plane is not following pitch as close as it should.

**Comment by kd0aij** [comment]
Thanks Lorenz. I appreciate the advice on tuning. No doubt the integrators are off by default and since the plane "felt" OK in the air I hadn't yet thought about it. 

I'll clean up the code and make a final commit. 

BTW, it may have been the low temp at the field today, but I had to recalibrate the airspeed sensor (reading 13m/sec on the ground with pitot covered). Offset changed from ~70 to ~50. 
And that was the new 3DR GPS module on this flight. Runway elevation is 1807m according to GE.

**Comment by kd0aij** [comment]
This PR is changing the sampling rate of the baro temp to a little less than 1Hz:
baro_temp_1hz

I'll check to see if there are any subscribers to that signal that might be needing/expecting the higher rate.

**Comment by LorenzMeier** [comment]
Ah yes - the multicopter code wants faster.

**Comment by kd0aij** [comment]
My code search failed to find any computational usages of baro_report.temperature or sensor_combined_s.baro_temp_celcius.  (not to say I didn't miss anything... used both a C index search and greplike search in Eclipse)

It looks like the sensor temp is used only in ms5611.cpp and the bandwidth of temperature should be sub-hertz, shouldn't it? The reduction in temperature sample frequency should introduce very little transient error in the typical case. Worst case would be for a scenario like the plot above, where the die temperature drops 8C over 60 seconds after takeoff.

On the other hand, the fidelity of pressure sampling is improved, since a sample is now missing only about once per second.

**Comment by kd0aij** [comment]
Found a problem with the low-freq temp sampling: choppy behavior of calculated pressure altitude.
Pressure altitude is very sensitive to air temperature, new question is whether one or both of temp and pressure altitude should be filtered.
temp_problem

**Comment by LorenzMeier** [comment]
@kd0aij The issue with filtering this is that we would need to reduce the bandwidth of the total output considerably to overcome this, right?
Maybe we should go back to the starting point of all this: Assuming you factor in the dT of all measurements, can we just filter correctly in the EKF? Because the varying intervals are not an issue per se theoretically (as far as I know), but rather an inconvenience for the post-filter routine.

**Comment by kd0aij** [comment]
Lorenz, I was actually wondering whether temp should be filtered to reduce noise on pressure altitude, given that the calculation is very sensitive to temp. But, assuming noise on the temp measurement is not significant, then the best thing to do is put ms5611.cpp back the way it was. I agree that the current filter implementation should handle irregular sampling correctly and bench testing today looks good with the original ms5611 sampling parameters.

So I'll revert to the original ms5611.cpp and remove the debug code. That will reduce the relevant changes to 1 file: ekf_att_pos_estimator.cpp. The change to LowPassFilter2p.cpp is no longer relevant to this PR, but does fix a bug in the reset method.

**Comment by kd0aij** [comment]
It would be nice to see some more flight testing on this, now that I think it's ready to merge.
Actually, this change affects all platforms using ekf_att_pos_estimator; doesn't that now include multirotors also?

**Comment by kd0aij** [comment]
@DrTon Hey, I was just reading up on the "moving average exponential filter" and found this on Wikipedia: http://en.wikipedia.org/wiki/Low-pass_filter#Simple_infinite_impulse_response_filter

``` 
// Return RC low-pass filter output samples, given input samples,
 // time interval dt, and time constant RC
 function lowpass(real[0..n] x, real dt, real RC)
   var real[0..n] y
   var real α := dt / (RC + dt)
   y[0] := x[0]
   for i from 1 to n
       y[i] := α * x[i] + (1-α) * y[i-1]
   return y
```
Looks like a more efficient implementation than using alpha = Math.expf(-dt/RC), but I don't see how to derive the formula for alpha.

**Comment by DrTon** [comment]
It's just an approximation of `exp` for `dt / RC ~ 0` :) Do you remember approximation of `exp`?: `exp(x) = 1 + x + o(x)` for `x ~ 0`. I.e. for small `dt / RC` this will give almost the same result, so use it if you are sure that it's always small, but for large `dt / RC` (general case) it will be not accurate.

**Comment by kd0aij** [comment]
:smile: I did the arithmetic wrong the first few times, as usual. Should just give up and stay current with Maple instead.
By my calculations the error in alpha is < 3% for RC > 4 dt, so it's fine for cutoff frequency < Nyquist/2

**Comment by kd0aij** [comment]
I'll try to get my PX4FMU/IO v1 quadcopter flying again to do some more flight testing of this branch

**Comment by sjwilks** [comment]
Hi Mark

I attempted to fly this yesterday but when I caught by another issue when I
was ready to test fly. That's resolved now so my next attempt should be in
the next day or two once the weather allows. I'm using a PX4FMU/IO setup.

Do you have any specific requests for what you want tested?

Simon

**Comment by kd0aij** [comment]
I think we just want to show that GPOS.Alt tracks closer to actual altitude with the dynamic value for _baro_gps_offset. This will be useful if the barometric pressure at "home" changes during the flight, and the filtered GPS altitude is an accurate reference.

**Comment by sjwilks** [comment]
I just did a short test flight. Logs and plots are here: http://dash.oznet.ch/view/cc4QP6t2SyEWXRaynkZG6A

**Comment by kd0aij** [comment]
Thanks Simon. The filtered baro and gps altitudes look good to me, as does the calculated offset. I don't understand the initial 5m offset between GPOS.Alt and GPS.Alt though. In my test, they started and ended with very little offset.
sjwilkstest1

**Comment by kd0aij** [comment]
I think the problem may be that _baro_ref was set before a start-up transient had settled out. Since we're not logging data till armed, these logfiles don't contain the necessary info to answer that question. It may be necessary to delay setting of _baro_ref until we have a stable reading of home altitude.

**Comment by LorenzMeier** [comment]
so is this good to merge?

**Comment by sjwilks** [comment]
From my point of view it's fine to merge as it certainly doesn't make
anything worse and seems to be heading in the right direction. I'll still
do another test flight though this time with logging turned on at system
startup time rather than just when arming. If there is still an offset it
could be addressed in a followup PR?

**Comment by LorenzMeier** [comment]
Makes sense, lets get it in!

**Comment by kd0aij** [comment]
agreed. Will do another flight test here once the wind gets down again.

**Comment by sjwilks** [comment]
I did one more test just now and there still seems to be an offset between GPOS.Alt and GPS.Alt of approx. 4-5m initially.

baro_alt

Logging started on boot rather than arming. The full log is here: http://dash.oznet.ch/logmuncher/charts/dJwj3x3vcfu6PYTvJ7Ea58/raw_flight_log.px4log

Otherwise it's looking good :)

**Comment by kd0aij** [comment]
@sjwilks This last log is interesting: between 10:55 and 10:57, baro alt increases, while gps alt decreases, and the offset goes from -18 to -10 (offset by 360 in this plot). What was happening at that time?
Also, the only thing I can see that would be responsible for the persistent offset between GPS and GPOS alt is _baro_ref_offset, but it's not logged and I don't understand how it's computed (ekf->states[9]).

offsettest2

**Comment by sjwilks** [comment]
Up until then I was in POSCTL and just going up and down at the same spot.
After that I landed and then flew around relatively quickly. It could be
that air pressure changes from the fast flight changed as while I have some
foam padding around the PX4 boards it only helps a little.

**Comment by kd0aij** [comment]
That rise may well be a weather offset :+1: While it does coincide with the start of your fast flying, it didn't go away during the minute from 10:57 to 10:58 when you weren't going fast. (velocities are offset by 340)
I wonder whether there might be some weather data available for your flight location to confirm that?

alt_vel

File: src/modules/ekf_att_pos_estimator/ekf_att_pos_estimator_main.cpp [hunk 1]
```diff
@@ -1195,7 +1225,11 @@ FixedwingEstimator::task_main()
 					// Set up height correctly
 					orb_copy(ORB_ID(sensor_baro0), _baro_sub, &_baro);
 					_baro_ref_offset = _ekf->states[9]; // this should become zero in the local frame
```
Comment (kd0aij) [inline src/modules/ekf_att_pos_estimator/ekf_att_pos_estimator_main.cpp]: What should be the value of states[9] when this is called? It looks like this is the only assignment to _baro_ref_offset.

**Comment by kd0aij** [comment]
@LorenzMeier It looks to me as though _baro_ref_offset is unnecessary or should be set to zero. See the line note above.

**Comment by kd0aij** [comment]
@DrTon Hey Anton, is there a way to plot the sum and difference of 2 signals in Flightplot?

**Comment by DrTon** [comment]
Unfortunately no :( If you have any idea how to implement parameters for such expressions I can try to implement this. As workaround you can easily add your own processor.

**Pull request merged**